In [1]:
## Handling  Imbalance Dataset.

## 1. Undersampling
## 2. Oversampling
## 3. SMOTETomek
## 4. Ensemble Techniques

In [2]:
import pandas as pd
df=pd.read_csv('Dataset/creditcard.csv')
df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [3]:
df.shape

(284807, 31)

In [4]:
df.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [5]:
# dependednt feature output value
df['Class'].value_counts()

0    284315
1       492
Name: Class, dtype: int64

* 0 are 284315 and 1 is 492, this is a very huge diffrence

In [6]:
# independent and dependent features
X = df.drop("Class", axis =1)
y = df.Class

## Cross Validation Like KFOLD and Hyperparameter tuning - first technique should consider when dealing with imbalance dataset

In [7]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
from sklearn.model_selection import KFold
import numpy as np
from sklearn.model_selection import GridSearchCV

In [8]:
log_class = LogisticRegression()
grid = {'C': 10.0 **np.arange(-2,3), 'penalty' : ['l1','l2']}
cv = KFold(n_splits = 5, random_state = None, shuffle = False)

In [9]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y, train_size = 0.7)

In [10]:
clf = GridSearchCV(log_class, grid, cv=cv, n_jobs = -1, scoring = 'f1_macro')
clf.fit(X_train,y_train)

C:\Users\sanjay\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py:372: FitFailedWarning: 
25 fits failed out of a total of 50.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
25 fits failed with the following error:
Traceback (most recent call last):
  File "C:\Users\sanjay\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py", line 680, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\sanjay\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py", line 1461, in fit
    solver = _check_solver(self.solver, self.penalty, self.dual)
  File "C:\Users\sanjay\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py", line 447, in _check_solver
    raise ValueError(


GridSearchCV(cv=KFold(n_splits=5, random_state=None, shuffle=False),
             estimator=LogisticRegression(), n_jobs=-1,
             param_grid={'C': array([1.e-02, 1.e-01, 1.e+00, 1.e+01, 1.e+02]),
                         'penalty': ['l1', 'l2']},
             scoring='f1_macro')

In [14]:
y_pred =  clf.predict(X_test)
print('-'*50)
print(confusion_matrix(y_test, y_pred))
print('-'*50)
print(accuracy_score(y_test, y_pred))
print('-'*50)
print(classification_report(y_test,y_pred))

--------------------------------------------------
[[85247    53]
 [   42   101]]
--------------------------------------------------
0.9988881476539916
--------------------------------------------------
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     85300
           1       0.66      0.71      0.68       143

    accuracy                           1.00     85443
   macro avg       0.83      0.85      0.84     85443
weighted avg       1.00      1.00      1.00     85443



In [15]:
y_train.value_counts()

0    199015
1       349
Name: Class, dtype: int64

In [16]:
from sklearn.ensemble import RandomForestClassifier
classifier = RandomForestClassifier()
classifier.fit(X_train,y_train)

RandomForestClassifier()

In [17]:
y_pred =  classifier.predict(X_test)
print('-'*50)
print(confusion_matrix(y_test, y_pred))
print('-'*50)
print(accuracy_score(y_test, y_pred))
print('-'*50)
print(classification_report(y_test,y_pred))

--------------------------------------------------
[[85291     9]
 [   32   111]]
--------------------------------------------------
0.9995201479348805
--------------------------------------------------
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     85300
           1       0.93      0.78      0.84       143

    accuracy                           1.00     85443
   macro avg       0.96      0.89      0.92     85443
weighted avg       1.00      1.00      1.00     85443



### using class weight

In [22]:
class_weight = dict({0:1,1:100}) # giving more importance to 1
#- loss of data so not preferable 

In [19]:
from sklearn.ensemble import RandomForestClassifier
classifier = RandomForestClassifier(class_weight= class_weight)
classifier.fit(X_train,y_train)

RandomForestClassifier(class_weight={0: 1, 1: 100})

In [20]:
y_pred =  classifier.predict(X_test)
print('-'*50)
print(confusion_matrix(y_test, y_pred))
print('-'*50)
print(accuracy_score(y_test, y_pred))
print('-'*50)
print(classification_report(y_test,y_pred))

--------------------------------------------------
[[85295     5]
 [   34   109]]
--------------------------------------------------
0.9995435553526912
--------------------------------------------------
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     85300
           1       0.96      0.76      0.85       143

    accuracy                           1.00     85443
   macro avg       0.98      0.88      0.92     85443
weighted avg       1.00      1.00      1.00     85443



## Undersampling

In [23]:
y_train.value_counts()

0    199015
1       349
Name: Class, dtype: int64

In [13]:
from collections import Counter
from imblearn.under_sampling import NearMiss

In [17]:
ns = NearMiss(sampling_strategy =0.8)
X_train_ns, y_train_ns = ns.fit_resample(X_train,y_train)
print("The no class before fit {}". format(Counter(y_train)))
print("The no class after fit {}". format(Counter(y_train_ns))) # counter --> like value_counts()

The no class before fit Counter({0: 199023, 1: 341})
The no class after fit Counter({0: 426, 1: 341})


#### here after resampling the majority class is reduced in such a way that 80% of majority class is the minority class


In [22]:
from sklearn.ensemble import RandomForestClassifier
classifier = RandomForestClassifier()
classifier.fit(X_train_ns,y_train_ns)

RandomForestClassifier()

In [23]:
y_pred =  classifier.predict(X_test)
print('-'*50)
print(confusion_matrix(y_test, y_pred))
print('-'*50)
print(accuracy_score(y_test, y_pred))
print('-'*50)
print(classification_report(y_test,y_pred))

--------------------------------------------------
[[70353 14939]
 [   11   140]]
--------------------------------------------------
0.825029551864986
--------------------------------------------------
              precision    recall  f1-score   support

           0       1.00      0.82      0.90     85292
           1       0.01      0.93      0.02       151

    accuracy                           0.83     85443
   macro avg       0.50      0.88      0.46     85443
weighted avg       1.00      0.83      0.90     85443



## Over Sampling

In [24]:
from imblearn.over_sampling import RandomOverSampler

In [26]:
os =  RandomOverSampler(sampling_strategy =0.5)
X_train_ns, y_train_ns = os.fit_resample(X_train,y_train)
print("The no class before fit {}". format(Counter(y_train)))
print("The no class after fit {}". format(Counter(y_train_ns))) # counter --> like value_counts()

The no class before fit Counter({0: 199023, 1: 341})
The no class after fit Counter({0: 199023, 1: 99511})


#### the minority class is increased and become equals to given percentage(our case 50%) of the majority class

In [27]:
from sklearn.ensemble import RandomForestClassifier
classifier = RandomForestClassifier()
classifier.fit(X_train_ns,y_train_ns)

RandomForestClassifier()

In [28]:
y_pred =  classifier.predict(X_test)
print('-'*50)
print(confusion_matrix(y_test, y_pred))
print('-'*50)
print(accuracy_score(y_test, y_pred))
print('-'*50)
print(classification_report(y_test,y_pred))

--------------------------------------------------
[[85287     5]
 [   30   121]]
--------------------------------------------------
0.9995903701883126
--------------------------------------------------
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     85292
           1       0.96      0.80      0.87       151

    accuracy                           1.00     85443
   macro avg       0.98      0.90      0.94     85443
weighted avg       1.00      1.00      1.00     85443



## SMOTETomek

In [34]:
from imblearn.combine import SMOTETomek # uses the combination of both undersampling and oversampling

In [37]:
# will take more time
os = SMOTETomek(sampling_strategy = 0.5)
X_train_ns, y_train_ns = os.fit_resample(X_train,y_train)
print("The number of classes before fit {}".format(Counter(y_train)))
print("The number of classes after fit {}".format(Counter(y_train_ns)))

KeyboardInterrupt: 

In [ ]:
from sklearn.ensemble import RandomForestClassifier
classifier=RandomForestClassifier()
classifier.fit(X_train_ns,y_train_ns)

In [ ]:
y_pred=classifier.predict(X_test)
print(confusion_matrix(y_test,y_pred))
print(accuracy_score(y_test,y_pred))
print(classification_report(y_test,y_pred))

## Ensemble Techniques

In [38]:
from imblearn.ensemble import EasyEnsembleClassifier

In [40]:
easy=EasyEnsembleClassifier()
easy.fit(X_train,y_train)

EasyEnsembleClassifier()

In [41]:
y_pred=easy.predict(X_test)
print(confusion_matrix(y_test,y_pred))
print(accuracy_score(y_test,y_pred))
print(classification_report(y_test,y_pred))

[[82674  2618]
 [   11   140]]
0.9692309492878293
              precision    recall  f1-score   support

           0       1.00      0.97      0.98     85292
           1       0.05      0.93      0.10       151

    accuracy                           0.97     85443
   macro avg       0.53      0.95      0.54     85443
weighted avg       1.00      0.97      0.98     85443

